### PR TITLE
Web UI

### PR DESCRIPTION
**Pull request by joelbryla**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -8,7 +8,8 @@
 from slack import Slack
 import cloudstorage as gcs
 
-app = Flask(__name__)
+app = Flask(__name__, static_url_path='')
```
Comment (auscompgeek) [inline main.py]: ???

File: main.py [hunk 2]
```diff
@@ -24,13 +25,17 @@
...
+
 @app.route('/')
 def hello():
-    return 'Hello World!'
+    return app.send_static_file('index.html')
 
 
 @app.route('/tba-webhook', methods=['POST'])
 def tba_webhook():
+    print("prediction")
```
Comment (auscompgeek) [inline main.py]: Please avoid using `print()`. If you wish to log things, use `app.logger`. I suspect this was for debugging though.

File: main.py [hunk 3]
```diff
@@ -44,6 +49,34 @@ def tba_webhook():
     return msg_type
 
 
+@app.route('/predictions')
+def give_predictions():
+    return str(trueskill_predictions)
```
Comment (auscompgeek) [inline main.py]: Please do not expose private team information.

File: main.py [hunk 4]
```diff
@@ -44,6 +49,34 @@ def tba_webhook():
...
+
+
+@app.route('/team_trueskill/<team>')
+def team_trueskill(team):
+    return str(trueskill.skill(team))
```
Comment (auscompgeek) [inline main.py]: This doesn't return plain text, and is prone to errors due to the lack of input checking. Use the API I put in instead.

File: main.py [hunk 5]
```diff
@@ -44,6 +49,34 @@ def tba_webhook():
...
+@app.route('/event_trueskill_ranking/<event>')
+def event_trueskil(event):
+    event_teams = trueskill.get_teams_at_event(event)
+    skills = [(team['key'], trueskill.skill(team['key']), team['nickname']) for team in event_teams]
+    skills = sorted(skills, key=lambda skill: skill[1], reverse=True)
+    msg = ''
+    num = 0
+    for skill in skills:
+        num += 1
+        msg += '<li>%s.  %s - %s - %0.1f\n</li> \n' % (str(num), skill[0][3:], skill[2], skill[1])
```
Comment (auscompgeek) [inline main.py]: I spy with my little eye an XSS vulnerability. Please do this properly.

File: static/main.js [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+				if (result.length === 0){
+					loaded = true
+					loading(false)
+					refresh()
+					return
+				}
+				for (team in result){
+					team_names[result[team].key] = result[team].nickname
+				}
+				get_team_names(page += 1)
```
Comment (auscompgeek) [inline static/main.js:426]: `+=`? You're not using this anywhere.

**Review by auscompgeek — CHANGES_REQUESTED** [review]
This otherwise looks good. Could you please update this?

File: main.py (hunk 3 above)
Comment (Ninjakow) [inline main.py]: What do you mean there is no slack key?

Comment (auscompgeek) [inline main.py]: Look, I don't understand why this is here, or why this needs to be here, or why you feel the need to expose our Slack data to those who don't have access to our Slack.

File: main.py (hunk 5 above)
Comment (auscompgeek) [reply main.py]: To clarify: generating HTML using format strings/string concatenation without ensuring everything is HTML escaped is a huge no-no.

Either use a templating engine that does HTML escaping, or return a JSON blob and do this client-side (by manipulating the DOM properly).

Comment (auscompgeek) [reply main.py]: In fact, didn't I put in an API that returns a JSON blob for this?

File: main.py [hunk 7]
```diff
@@ -44,6 +49,34 @@ def tba_webhook():
...
+    return msg.encode('utf-8').strip()
+
+
+@app.route('/predict')
+def send_predict_page():
+    return app.send_static_file('predict.html')
```
Comment (auscompgeek) [inline main.py:61]: Why are you calling an internal Flask method?

File: main.py (hunk 3 above)
Comment (Ninjakow) [inline main.py]: How does it expose our slack data?

**Comment by auscompgeek** [comment]
Oh boy. I just opened `static/main.js` in my editor. Not only does it break my editor('s syntax highlighting), it also breaks ESLint, and even JSHint didn't catch a bug I just found (after fixing the file enough so that JSHint actually runs). This is going to be a long night…

**Comment by auscompgeek** [comment]
Someone remind me to deploy to App Engine in the morning…

**Pull request merged**